// @openfga/sdk/dist/configuration.js
"use strict";
/**
 * JavaScript and Node.js SDK for OpenFGA
 *
 * API version: 0.1
 * Website: https://openfga.dev
 * Documentation: https://openfga.dev/docs
 * Support: https://openfga.dev/community
 * License: [Apache-2.0](https://github.com/openfga/js-sdk/blob/main/LICENSE)
 *
 * NOTE: This file was auto generated by OpenAPI Generator (https://openapi-generator.tech). DO NOT EDIT.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Configuration = exports.GetDefaultRetryParams = void 0;
const types_1 = require("./credentials/types");
const errors_1 = require("./errors");
const validation_1 = require("./validation");
// default maximum number of retry
const DEFAULT_MAX_RETRY = 15;
// default minimum wait period in retry - but will backoff exponentially
const DEFAULT_MIN_WAIT_MS = 100;
const DEFAULT_USER_AGENT = "openfga-sdk js/0.3.5";
function GetDefaultRetryParams(maxRetry = DEFAULT_MAX_RETRY, minWaitInMs = DEFAULT_MIN_WAIT_MS) {
    return {
        maxRetry: maxRetry,
        minWaitInMs: minWaitInMs,
    };
}
exports.GetDefaultRetryParams = GetDefaultRetryParams;
class Configuration {
    constructor(params = {}) {
        /**
         * provide scheme (e.g. `https`)
         *
         * @type {string}
         * @memberof Configuration
         * @deprecated
         */
        this.apiScheme = "https";
        /**
         * Returns the API base path (apiScheme+apiHost)
         */
        this.getBasePath = () => {
            if (this.apiUrl) {
                return this.apiUrl;
            }
            else {
                return `${this.apiScheme}://${this.apiHost}`;
            }
        };
        this.apiScheme = params.apiScheme || this.apiScheme;
        this.apiHost = params.apiHost;
        this.apiUrl = params.apiUrl;
        this.storeId = params.storeId;
        const credentialParams = params.credentials;
        if (credentialParams) {
            switch (credentialParams === null || credentialParams === void 0 ? void 0 : credentialParams.method) {
                case types_1.CredentialsMethod.ApiToken:
                    this.credentials = {
                        method: credentialParams.method,
                        config: {
                            token: credentialParams.config.token,
                            headerName: "Authorization",
                            headerValuePrefix: "Bearer",
                        }
                    };
                    break;
                case types_1.CredentialsMethod.ClientCredentials:
                    this.credentials = {
                        method: types_1.CredentialsMethod.ClientCredentials,
                        config: {
                            // We are only copying them from the passed in params here. We will be validating that they are valid in the Credentials constructor
                            clientId: credentialParams.config.clientId,
                            clientSecret: credentialParams.config.clientSecret,
                            apiAudience: credentialParams.config.apiAudience,
                            apiTokenIssuer: credentialParams.config.apiTokenIssuer,
                        }
                    };
                    break;
                case types_1.CredentialsMethod.None:
                default:
                    this.credentials = { method: types_1.CredentialsMethod.None };
                    break;
            }
        }
        const baseOptions = params.baseOptions || {};
        baseOptions.headers = baseOptions.headers || {};
        if (typeof process === "object" && process.title === "node" && !baseOptions.headers["User-Agent"]) {
            baseOptions.headers["User-Agent"] = DEFAULT_USER_AGENT;
        }
        this.baseOptions = baseOptions;
        this.retryParams = params.retryParams;
    }
    /**
     *
     * @return {boolean}
     * @throws {FgaValidationError}
     */
    isValid() {
        var _a;
        if (!this.apiUrl) {
            (0, validation_1.assertParamExists)("Configuration", "apiScheme", this.apiScheme);
            (0, validation_1.assertParamExists)("Configuration", "apiHost", this.apiHost);
        }
        if (!(0, validation_1.isWellFormedUriString)(this.getBasePath())) {
            throw new errors_1.FgaValidationError(this.apiUrl ?
                `Configuration.apiUrl (${this.apiUrl}) is not a valid URI (${this.getBasePath()})` :
                `Configuration.apiScheme (${this.apiScheme}) and Configuration.apiHost (${this.apiHost}) do not form a valid URI (${this.getBasePath()})`);
        }
        if (this.storeId && !(0, validation_1.isWellFormedUlidString)(this.storeId)) {
            throw new errors_1.FgaValidationError("storeId");
        }
        if (((_a = this.retryParams) === null || _a === void 0 ? void 0 : _a.maxRetry) && this.retryParams.maxRetry > 15) {
            throw new errors_1.FgaValidationError("Configuration.retryParams.maxRetry exceeds maximum allowed limit of 15");
        }
        return true;
    }
}
exports.Configuration = Configuration;
/**
 * Defines the version of the SDK
 *
 * @private
 * @type {string}
 * @memberof Configuration
 */
Configuration.sdkVersion = "0.3.5";
