// @openfga/sdk/dist/errors.js
"use strict";
/**
 * JavaScript and Node.js SDK for OpenFGA
 *
 * API version: 0.1
 * Website: https://openfga.dev
 * Documentation: https://openfga.dev/docs
 * Support: https://openfga.dev/community
 * License: [Apache-2.0](https://github.com/openfga/js-sdk/blob/main/LICENSE)
 *
 * NOTE: This file was auto generated by OpenAPI Generator (https://openapi-generator.tech). DO NOT EDIT.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.FgaRequiredParamError = exports.FgaValidationError = exports.FgaApiAuthenticationError = exports.FgaApiInternalError = exports.FgaApiRateLimitExceededError = exports.FgaApiNotFoundError = exports.FgaApiValidationError = exports.FgaApiError = exports.FgaError = void 0;
/**
 *
 * @export
 * @class FgaError
 * @extends {Error}
 */
class FgaError extends Error {
    constructor(err, msg) {
        super(msg || typeof err === "string"
            ? err
            : `FGA Error${(err === null || err === void 0 ? void 0 : err.message) ? `: ${err.message}` : ""}`);
        this.name = "FgaError";
        if (err === null || err === void 0 ? void 0 : err.stack) {
            this.stack = err.stack;
        }
    }
}
exports.FgaError = FgaError;
function getRequestMetadataFromPath(path) {
    // This function works because all paths start with /stores/{storeId}/{type}
    let splitPath = (path || "").split("/");
    if (splitPath.length < 4) {
        splitPath = [];
    }
    const storeId = splitPath[2] || "";
    const endpointCategory = splitPath[3] || "";
    return { storeId, endpointCategory };
}
const cFGARequestId = "fga-request-id";
function getResponseHeaders(err) {
    return err.response
        ? Object.fromEntries(Object.entries(err.response.headers).map(([k, v]) => [
            k.toLowerCase(), v,
        ]))
        : {};
}
/**
 *
 * @export
 * @class FgaApiError
 * @extends { FgaError }
 */
class FgaApiError extends FgaError {
    constructor(err, msg) {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k;
        super(msg ? msg : err);
        this.name = "FgaApiError";
        this.statusCode = (_a = err.response) === null || _a === void 0 ? void 0 : _a.status;
        this.statusText = (_b = err.response) === null || _b === void 0 ? void 0 : _b.statusText;
        this.requestData = (_c = err.config) === null || _c === void 0 ? void 0 : _c.data;
        this.requestURL = (_d = err.config) === null || _d === void 0 ? void 0 : _d.url;
        this.method = (_e = err.config) === null || _e === void 0 ? void 0 : _e.method;
        const { storeId, endpointCategory } = getRequestMetadataFromPath((_f = err.request) === null || _f === void 0 ? void 0 : _f.path);
        this.storeId = storeId;
        this.endpointCategory = endpointCategory;
        this.apiErrorMessage = (_h = (_g = err.response) === null || _g === void 0 ? void 0 : _g.data) === null || _h === void 0 ? void 0 : _h.message;
        this.responseData = (_j = err.response) === null || _j === void 0 ? void 0 : _j.data;
        this.responseHeader = (_k = err.response) === null || _k === void 0 ? void 0 : _k.headers;
        const errResponseHeaders = getResponseHeaders(err);
        this.requestId = errResponseHeaders[cFGARequestId];
        if (err === null || err === void 0 ? void 0 : err.stack) {
            this.stack = err.stack;
        }
    }
}
exports.FgaApiError = FgaApiError;
/**
 *
 * @export
 * @class FgaApiValidationError
 * @extends { FgaApiError }
 */
class FgaApiValidationError extends FgaApiError {
    constructor(err, msg) {
        var _a, _b, _c, _d, _e, _f, _g, _h;
        // If there is a better error message, use it instead of the default error
        super(err);
        this.name = "FgaApiValidationError";
        this.apiErrorCode = (_b = (_a = err.response) === null || _a === void 0 ? void 0 : _a.data) === null || _b === void 0 ? void 0 : _b.code;
        const { endpointCategory } = getRequestMetadataFromPath((_c = err.request) === null || _c === void 0 ? void 0 : _c.path);
        this.message = msg
            ? msg
            : ((_e = (_d = err.response) === null || _d === void 0 ? void 0 : _d.data) === null || _e === void 0 ? void 0 : _e.message)
                ? `FGA API Validation Error: ${(_f = err.config) === null || _f === void 0 ? void 0 : _f.method} ${endpointCategory} : Error ${(_h = (_g = err.response) === null || _g === void 0 ? void 0 : _g.data) === null || _h === void 0 ? void 0 : _h.message}`
                : err.message;
    }
}
exports.FgaApiValidationError = FgaApiValidationError;
/**
 *
 * @export
 * @class FgaApiNotFoundError
 * @extends { FgaApiError }
 */
class FgaApiNotFoundError extends FgaApiError {
    constructor(err, msg) {
        var _a, _b, _c, _d, _e, _f, _g;
        // If there is a better error message, use it instead of the default error
        super(err);
        this.name = "FgaApiNotFoundError";
        this.apiErrorCode = (_b = (_a = err.response) === null || _a === void 0 ? void 0 : _a.data) === null || _b === void 0 ? void 0 : _b.code;
        this.message = msg
            ? msg
            : ((_d = (_c = err.response) === null || _c === void 0 ? void 0 : _c.data) === null || _d === void 0 ? void 0 : _d.message)
                ? `FGA API NotFound Error: ${(_e = err.config) === null || _e === void 0 ? void 0 : _e.method} : Error ${(_g = (_f = err.response) === null || _f === void 0 ? void 0 : _f.data) === null || _g === void 0 ? void 0 : _g.message}`
                : err.message;
    }
}
exports.FgaApiNotFoundError = FgaApiNotFoundError;
const cXRateLimit = "x-ratelimit-limit";
const cXRateLimitReset = "x-ratelimit-reset";
/**
 *
 * @export
 * @class FgaApiRateLimitExceededError
 * @extends { FgaApiError }
 */
class FgaApiRateLimitExceededError extends FgaApiError {
    constructor(err, msg) {
        var _a, _b, _c;
        super(err);
        this.name = "FgaApiRateLimitExceededError";
        this.apiErrorCode = (_b = (_a = err.response) === null || _a === void 0 ? void 0 : _a.data) === null || _b === void 0 ? void 0 : _b.code;
        const { endpointCategory } = getRequestMetadataFromPath((_c = err.request) === null || _c === void 0 ? void 0 : _c.path);
        const errResponseHeaders = getResponseHeaders(err);
        this.message = msg
            ? msg
            : `FGA API Rate Limit Error for ${this.method} ${endpointCategory}`;
    }
}
exports.FgaApiRateLimitExceededError = FgaApiRateLimitExceededError;
/**
 *
 * @export
 * @class FgaApiInternalError
 * @extends { FgaApiError }
 */
class FgaApiInternalError extends FgaApiError {
    constructor(err, msg) {
        var _a, _b, _c, _d, _e, _f, _g, _h;
        // If there is a better error message, use it instead of the default error
        super(err);
        this.name = "FgaApiInternalError";
        const { endpointCategory } = getRequestMetadataFromPath((_a = err.request) === null || _a === void 0 ? void 0 : _a.path);
        this.apiErrorCode = (_c = (_b = err.response) === null || _b === void 0 ? void 0 : _b.data) === null || _c === void 0 ? void 0 : _c.code;
        this.message = msg
            ? msg
            : ((_e = (_d = err.response) === null || _d === void 0 ? void 0 : _d.data) === null || _e === void 0 ? void 0 : _e.message)
                ? `FGA API Internal Error: ${(_f = err.config) === null || _f === void 0 ? void 0 : _f.method} ${endpointCategory} : Error ${(_h = (_g = err.response) === null || _g === void 0 ? void 0 : _g.data) === null || _h === void 0 ? void 0 : _h.message}`
                : err.message;
    }
}
exports.FgaApiInternalError = FgaApiInternalError;
/**
 *
 * @export
 * @class FgaApiAuthenticationError
 * @extends { FgaApiError }
 */
class FgaApiAuthenticationError extends FgaError {
    constructor(err) {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k;
        super(`FGA Authentication Error.${((_a = err.response) === null || _a === void 0 ? void 0 : _a.statusText) ? ` ${err.response.statusText}` : ""}`);
        this.name = "FgaApiAuthenticationError";
        this.statusCode = (_b = err.response) === null || _b === void 0 ? void 0 : _b.status;
        this.statusText = (_c = err.response) === null || _c === void 0 ? void 0 : _c.statusText;
        this.requestURL = (_d = err.config) === null || _d === void 0 ? void 0 : _d.url;
        this.method = (_e = err.config) === null || _e === void 0 ? void 0 : _e.method;
        this.responseData = (_f = err.response) === null || _f === void 0 ? void 0 : _f.data;
        this.responseHeader = (_g = err.response) === null || _g === void 0 ? void 0 : _g.headers;
        this.apiErrorCode = (_j = (_h = err.response) === null || _h === void 0 ? void 0 : _h.data) === null || _j === void 0 ? void 0 : _j.code;
        const errResponseHeaders = getResponseHeaders(err);
        this.requestId = errResponseHeaders[cFGARequestId];
        let data;
        try {
            data = JSON.parse(((_k = err.config) === null || _k === void 0 ? void 0 : _k.data) || "{}");
        }
        catch (err) {
            /* do nothing */
        }
        this.clientId = data === null || data === void 0 ? void 0 : data.client_id;
        this.audience = data === null || data === void 0 ? void 0 : data.audience;
        this.grantType = data === null || data === void 0 ? void 0 : data.grant_type;
        if (err === null || err === void 0 ? void 0 : err.stack) {
            this.stack = err.stack;
        }
    }
}
exports.FgaApiAuthenticationError = FgaApiAuthenticationError;
/**
 *
 * @export
 * @class FgaValidationError
 * @extends { FgaError }
 */
class FgaValidationError extends FgaError {
    constructor(field, msg) {
        super(msg);
        this.field = field;
        this.name = "FgaValidationError";
    }
}
exports.FgaValidationError = FgaValidationError;
/**
 *
 * @export
 * @class FgaRequiredParamError
 * @extends { FgaValidationError }
 */
class FgaRequiredParamError extends FgaValidationError {
    constructor(functionName, field, msg) {
        super(field, msg);
        this.functionName = functionName;
        this.name = "FgaRequiredParamError";
    }
}
exports.FgaRequiredParamError = FgaRequiredParamError;
